### PR TITLE
Ellipsoid geometry fixes

### PR DESCRIPTION
**Pull request by MFraters**
This commit contains a large rework of the ellipsoidal code and quite some bug fixes. It is now connected to the deal.ii geometry (subdivided) parallelepiped in stead of the (subdivided) hyper rectangle. This change should simplify the code and make it less error prone, because it uses a deal.ii intrinsic geometry. The tests are also changed to assign a specific temperature to each side of ellipsoid so the test also checks if the mapping of the sides is still correct.

### CODE REVIEW

File: include/aspect/geometry_model/ellipsoidal_chunk.h [hunk 1]
```diff
@@ -185,31 +171,10 @@ namespace aspect
...
-        /**
-         * Retrieve the rotation_longitude variable.
-         */
-        double
-        get_rotation_longitude() const;
-
-
-        /**
-         * Retrieve the rotation_latitude variable.
+         * Retrieve the value of the eccentricity.
```
Comment (gassmoeller) [inline include/aspect/geometry_model/ellipsoidal_chunk.h]: This comment is out of date considering the function name?

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 2]
```diff
@@ -154,80 +145,38 @@ namespace aspect
...
-          GridTools::transform (std_cxx11::bind(&EllipsoidalChunkGeometry::rotate_parallelogram,
-                                                std_cxx11::cref(manifold),
-                                                std_cxx11::_1),
-                                coarse_grid);
-        }
-      //transform to the ellipsoid surface
+
+      /**
+       * Generate parallelepiped grid with one point (point 0) at (0,0,0) and the
+       * other corners (respectivly corner 1,2 and 4) placed relative to that point.
```
Comment (gassmoeller) [inline source/geometry_model/ellipsoidal_chunk.cc]: respectively ... and corners 1, 2 and 3?

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 3]
```diff
@@ -154,80 +145,38 @@ namespace aspect
...
+       */
+      const Point<3> corner_points[dim] = {Point<dim>((corners[1][0]-corners[0][0])*numbers::PI/180,
+                                                      (corners[1][1]-corners[0][1])*numbers::PI/180,
+                                                      0),
+                                           Point<dim>((corners[3][0]-corners[0][0])*numbers::PI/180,
+                                                      (corners[3][1]-corners[0][1])*numbers::PI/180,
```
Comment (gassmoeller) [inline source/geometry_model/ellipsoidal_chunk.cc]: should this be corners[3] - corners[2]? Or does it not matter, because corners[0] and corners[2] are on the same latitude/longitude?

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 4]
```diff
@@ -301,12 +250,12 @@ namespace aspect
...
-                  std::pair<std::string,types::boundary_id>("west",  1),
-                  std::pair<std::string,types::boundary_id>("south",  2),
-                  std::pair<std::string,types::boundary_id>("north",   3),
+              = { std::pair<std::string,types::boundary_id>("west", 0),
+                  std::pair<std::string,types::boundary_id>("east", 1),
+                  std::pair<std::string,types::boundary_id>("north",  2),
+                  std::pair<std::string,types::boundary_id>("south",  3),
                   std::pair<std::string,types::boundary_id>("inner", 4),
-                  std::pair<std::string,types::boundary_id>("outer",    5)
+                  std::pair<std::string,types::boundary_id>("outer", 5)
```
Comment (gassmoeller) [inline source/geometry_model/ellipsoidal_chunk.cc]: please align the spaces again (also in the lines above)

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 5]
```diff
@@ -592,8 +491,20 @@ namespace aspect
                   corners[1][0] = corners[2][0];
                   corners[3][0] = corners[0][0];
                 }
-              corners_rectangle = corners;
             }
+          // Check that the calculated corners also obey the rules for the location of the coners.
```
Comment (gassmoeller) [inline source/geometry_model/ellipsoidal_chunk.cc]: corners

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 6]
```diff
@@ -592,8 +491,20 @@ namespace aspect
...
+
+          AssertThrow (corners[0][1] >= corners[3][1],
+                       ExcMessage ("The latitude of the NE (" + boost::lexical_cast<std::string>(corners[0][1]) + ") corner cannot be larger then the longitude of the SE corner (" + boost::lexical_cast<std::string>(corners[3][1]) + "). This is an internal check, if you see this please contact the developer."));
+
+          AssertThrow (corners[2][0] <= corners[3][0],
+                       ExcMessage ("The longitude of the SW (" + boost::lexical_cast<std::string>(corners[2][0]) + ") corner cannot be larger then the longitude of the SE corner (" + boost::lexical_cast<std::string>(corners[3][0]) + "). This is an internal check, if you see this please contact the developer."));
+
+          AssertThrow (corners[2][1] <= corners[1][1],
+                       ExcMessage ("The latitude of the SW corner (" + boost::lexical_cast<std::string>(corners[2][1]) + ") cannot be smaller then the longitude of the NW corner (" + boost::lexical_cast<std::string>(corners[1][1]) + "). This is an internal check, if you see this please contact the developer."));
```
Comment (gassmoeller) [inline source/geometry_model/ellipsoidal_chunk.cc]: If these are internal checks that should only fail if the above formula is wrong or somebody fiddles with the code, then they should be Assert instead of AssertThrow. Not very important in this case though.

File: tests/ellipsoidal_chunk_coordinate_parallel.prm [hunk 7]
```diff
@@ -19,8 +19,8 @@ set End time                               = 0
 subsection Geometry model
   set Model name = ellipsoidal chunk
     subsection Ellipsoidal chunk
-        set NE corner = -10:30
-	set SW corner = 20:-10
+        set NE corner = 20:30
+	set SW corner = -10:-10
```
Comment (gassmoeller) [inline tests/ellipsoidal_chunk_coordinate_parallel.prm]: please align the parameters

File: tests/ellipsoidal_chunk_noncoordinate_parallel.prm [hunk 8]
```diff
@@ -19,8 +19,8 @@ set End time                               = 0
 subsection Geometry model
   set Model name = ellipsoidal chunk
     subsection Ellipsoidal chunk
-        set NE corner = -10:30
-	set NW corner = 10:40
+        set NE corner = 10:40
```
Comment (gassmoeller) [inline tests/ellipsoidal_chunk_noncoordinate_parallel.prm]: alignment

File: tests/ellipsoidal_chunk_noncoordinate_parallel.prm [hunk 9]
```diff
@@ -36,17 +36,16 @@ subsection Model settings
   set Tangential velocity boundary indicators = outer,north,south,east,west
   set Prescribed velocity boundary indicators =
 
-  set Fixed temperature boundary indicators   = inner, outer
+  set Fixed temperature boundary indicators   = inner, outer, east, west, north,south
```
Comment (gassmoeller) [inline tests/ellipsoidal_chunk_noncoordinate_parallel.prm]: space before south ;-)

File: tests/ellipsoidal_chunk_noncoordinate_parallel.prm [hunk 10]
```diff
@@ -86,7 +85,7 @@ subsection Postprocess
 
   subsection Visualization
     set List of output variables 	  = density
-    set Output format                 = gnuplot
+    set Output format                     = gnuplot
```
Comment (gassmoeller) [inline tests/ellipsoidal_chunk_noncoordinate_parallel.prm]: Either do not change the line, or align it. Currently, it is misaligned by one ;-) (or is my browser strange?)

**Comment by gassmoeller** [comment]
Very nice simplification and improvements. I found only stylistic things and typos. I hope we do not need to update this test very often, thousands of lines of test output increase the size of the repository over time.

File: source/geometry_model/ellipsoidal_chunk.cc (hunk 2 above)
Comment (MFraters) [inline source/geometry_model/ellipsoidal_chunk.cc]: I am referring here to how the parallelepiped geometry refers to its own points. The three points you give are points 1,2 and 4. Point 0 is always at (0,0,0) (that's why I need to shift the whole grid later on), and points 3,5,6 and 7 are calculated internally.

File: source/geometry_model/ellipsoidal_chunk.cc (hunk 3 above)
Comment (MFraters) [inline source/geometry_model/ellipsoidal_chunk.cc]: Corner two is actually never used in this implementation. What I do here is calculate the difference from corner zero in each direction, and use those differences to create the parallelepiped.

File: source/geometry_model/ellipsoidal_chunk.cc (hunk 2 above)
Comment (gassmoeller) [inline source/geometry_model/ellipsoidal_chunk.cc]: Ah, you are right. I just looked for the symmetry of the numbers, but we are in 3d of course (8 corners), and only need 3 other points to define the parallelepiped.

File: source/geometry_model/ellipsoidal_chunk.cc (hunk 3 above)
Comment (gassmoeller) [inline source/geometry_model/ellipsoidal_chunk.cc]: Same as above. Thanks for the explanation.

**Comment by MFraters** [comment]
Hey Rene,
Thanks for the comments! I have addressed most of them, but the I think I see the allign things in the prm files different then you do somehow, because both in vi and here on github they seems to be alligned to me :s

**Comment by gassmoeller** [comment]
Hi Menno,
dont worry, then we will simply merge it like this. Some of these cases look for me like github interprets the spaces as tabs ( e.g. tests/ellipsoidal_chunk_coordinate_parallel.prm:23). Could you check for a second if there are tabs instead of spaces in your file? If not, lets not bother and just blame my browser ;-)

**Comment by bangerth** [comment]
On 10/06/2015 07:35 AM, MFraters wrote:
> Thanks for the comments! I have addressed most of them, but the I think I see
> the allign things in the prm files different then you do somehow, because both
> in vi and here on github they seems to be alligned to me :s

There may be tabs in these files. If you convert tabs into spaces, everyone 
should see things the same.

**Comment by MFraters** [comment]
I think I fixed that spaces/tabs problem. Although it compiles fine on my machine, the tester has some problems with compiling this. The main problem seems to be that it wants to use the shift with a point instead of a tensor. This while the deal.ii manual explicitly states (http://dealii.org/developer/doxygen/deal.II/namespaceGridTools.html#a4b953611883979e3f920f800bd855df3) that a tensor should be used. I am using deal.ii 8.3.pre.

**Comment by bangerth** [comment]
The Tester is apparently using an older version of deal ii. Just pass a point instead of the tensor, this is backward compatible.

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 11]
```diff
@@ -53,11 +53,8 @@ namespace aspect
       semi_major_axis_a (-1),
       eccentricity (-1),
       semi_minor_axis_b (-1),
-      rot_para_to_para_angle (0),
-      para_to_rect_angle (0),
-      rotation_longitude (-1),
-      rotation_latitude (-1),
-      bottom_depth (-1)
+      bottom_depth (-1),
+      corners({{0}})
```
Comment (gassmoeller) [inline source/geometry_model/ellipsoidal_chunk.cc]: clang gives a warning in this line (too many braces for scalar initializer), I think you might as well use the default constructor corners(). If anybody tries to access non-existent entries they will get an error anyway.

**Comment by gassmoeller** [comment]
gcc throws floating point exceptions in your tests, but clang does not seem to care. If you have a debugger available and compiled Aspect with `ASPECT_USE_FP_EXCEPTIONS = ON` you could look for the place, otherwise we will need to look for a division-by-zero or something similar.

**Comment by MFraters** [comment]
Thanks for the tip! I was already wondering why my local gcc compiler had no problem with it, and the tester did.

File: source/geometry_model/ellipsoidal_chunk.cc (hunk 11 above)
Comment (bangerth) [inline source/geometry_model/ellipsoidal_chunk.cc]: You can't use this syntax, actually -- it requires support for C++11 which we don't assume yet everyone has.

Since `corners` is a `std::vector` and such objects have a default constructor that initializes it to be empty, my suggestion is to simply remove the entire line.

**Comment by MFraters** [comment]
Maybe I should make a separate issue on this, but when I turn this option on in debug mode it crashes with the following error, exactly the first time it encounters patterns::Double(), which is in parameters.cc on line 88-92:

```
[gaia:28259] *** Process received signal ***
[gaia:28259] Signal: Floating point exception (8)
[gaia:28259] Signal code: Invalid floating point operation (7)
[gaia:28259] Failing at address: 0x3f1ac41cb0
[gaia:28259] [ 0] /lib64/libpthread.so.0[0x3f1b00f710]
[gaia:28259] [ 1] /lib64/libc.so.6[0x3f1ac41cb0]
[gaia:28259] [ 2] /lib64/libc.so.6(__printf_fp+0x97e)[0x3f1ac4a53e]
[gaia:28259] [ 3] /lib64/libc.so.6(_IO_vfprintf+0x18d0)[0x3f1ac458a0]
[gaia:28259] [ 4] /lib64/libc.so.6(vsnprintf+0xa2)[0x3f1ac6f752]
[gaia:28259] [ 5] /usr/local/GCC/lib64/libstdc++.so.6(+0x88f2f)[0x7fffe391ff2f]
[gaia:28259] [ 6] /usr/local/GCC/lib64/libstdc++.so.6(_ZNKSt7num_putIcSt19ostreambuf_iteratorIcSt11char_traitsIcEEE15_M_insert_floatIdEES3_S3_RSt8ios_baseccT_+0xaa)[0x7fffe3926eca]
[gaia:28259] [ 7] /usr/local/GCC/lib64/libstdc++.so.6(_ZNKSt7num_putIcSt19ostreambuf_iteratorIcSt11char_traitsIcEEE6do_putES3_RSt8ios_basecd+0x10)[0x7fffe3927190]
[gaia:28259] [ 8] /usr/local/GCC/lib64/libstdc++.so.6(_ZNSo9_M_insertIdEERSoT_+0xc5)[0x7fffe3933b35]
[gaia:28259] [ 9] /home/aspect/deal.ii/deal.ii_2015-06-04-v11.4/dealii/build/lib/libdeal_II.g.so.8.3.pre(_ZNK6dealii8Patterns6Double11descriptionEv+0x238)[0x7ffff5e2ed46]
[gaia:28259] [10] /home/aspect/deal.ii/deal.ii_2015-06-04-v11.4/dealii/build/lib/libdeal_II.g.so.8.3.pre(_ZN6dealii16ParameterHandler13declare_entryERKSsS2_RKNS_8Patterns11PatternBaseES2_+0x625)[0x7ffff5e391ed]
[gaia:28259] [11] ./aspect(_ZN6aspect10ParametersILi2EE18declare_parametersERN6dealii16ParameterHandlerE+0x62a)[0x979f7e]
[gaia:28259] [12] ./aspect(_ZN6aspect9SimulatorILi2EE18declare_parametersERN6dealii16ParameterHandlerE+0x9)[0x97f2d9]
[gaia:28259] [13] ./aspect(main+0x440)[0x9efa7c]
[gaia:28259] [14] /lib64/libc.so.6(__libc_start_main+0xfd)[0x3f1ac1ed5d]
[gaia:28259] [15] ./aspect[0x643099]
[gaia:28259] *** End of error message ***
```
In release mode it gives me no problem at all. This suggests to me that somewhere in the declare_entry function an assert calling a function is causing the problem, although it does seem to try to insert something (which would be strange for an assert). I am trying to look into this, but suggestions are welcome :)

File: source/geometry_model/ellipsoidal_chunk.cc (hunk 11 above)
Comment (MFraters) [inline source/geometry_model/ellipsoidal_chunk.cc]: ok, I will remove it.

**Comment by gassmoeller** [comment]
I am not sure what is happening there. It seems like the parameter handler does not like to output the description , see line [9]. (i.e. the bounds, min max value of Patterns::Double() on your system). They should be `-std::numeric_limits<double>::max()` and `std::numeric_limits<double>::max()`. I have too little experience with floating point exceptions to say what might be the reason for this. What operating system and compiler are you using? 
It is unfortunate that this prevents us from finding the real problem.
I can reproduce your problem on my machine, but the reason it is a bit obscured. The exception gets thrown inside compute_nonzero_normal_flux_constraints (which is called from compute_no_normal_flux_constraints) in include/deal.II/numerics/vector_tools.templates.h:5546, because the normal vector of a face is (0,0,0) and therefore the norm is zero. Have you checked that the generated mesh is actually what you expect it to be, and that all corner cases are handled correctly (like theta, phi = 0 or pi)?

**Comment by bangerth** [comment]
On 10/07/2015 10:33 AM, Rene Gassmöller wrote:
> I am not sure what is happening there. It seems like the parameter
> handler does not like to output the description , see line [9]. (i.e.
> the bounds, min max value of Patterns::Double() on your system). They
> should be |-std::numeric_limits<double>::max()| and
> |std::numeric_limits<double>::max()|. I have too little experience with
> floating point exceptions to say what might be the reason for this. What
> operating system and compiler are you using?

Let us not try to dig into this here -- there is nothing Menno can do 
about it. There is discussion in #500 and #508 on this issue.

> It is unfortunate that this prevents us from finding the real problem.
> I can reproduce your problem on my machine, but the reason it is a bit
> obscured. The exception gets thrown inside
> compute_nonzero_normal_flux_constraints (which is called from
> compute_no_normal_flux_constraints) in
> include/deal.II/numerics/vector_tools.templates.h:5546, because the
> normal vector of a face is (0,0,0) and therefore the norm is zero. Have
> you checked that the generated mesh is actually what you expect it to
> be, and that all corner cases are handled correctly (like theta, phi = 0
> or pi)?

That seems like the problem. Certainly the normal vector should never be 
zero.

**Comment by tjhei** [comment]
Well, there is certainly a problem if the tester aborts with an FPE for these tests. Let me know if you need help with this, Menno.

**Comment by MFraters** [comment]
Thanks for all the comments! I checked many corner cases, including one in which a one corner is exactly at (0,0,0), and the meshes look exactly what I expect them to look like. 

The only reason I can think of that the normal vector is zero is that all the points of the face are at the same location. Is this correct? I also looked at the gnuplot and vtk files (through coarse_grid.write_mesh_vtk("mesh");). The gnuplot looks a bit odd, but I don't know the format that well. The vtk file looks fine, with no double points. To be sure I also looked at the minimal_cell_diameter, which also returned a reasonable value.

What could also be the case is that it has something to do with how I set the boundary indicator. I noticed with writing the new version of this plugin that I can't just name a side what I want or it would (if I remember correctly) give similar crashes, which I found a bit odd.

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 12]
```diff
@@ -453,12 +381,27 @@ namespace aspect
...
-          corners[0] = Utilities::string_to_double(Utilities::split_string_list(NEcorner,':'));
-          corners[1] = Utilities::string_to_double(Utilities::split_string_list(NWcorner,':'));
-          corners[2] = Utilities::string_to_double(Utilities::split_string_list(SWcorner,':'));
-          corners[3] = Utilities::string_to_double(Utilities::split_string_list(SEcorner,':'));
-          rotation_longitude = corners[0][0];
-          rotation_latitude = corners[0][1];
+          if (present[0])
+            corners[0] = Utilities::string_to_double(Utilities::split_string_list(NEcorner,':'));
+          else
+            corners[0]= {0,0,0};
```
Comment (tjhei) [inline source/geometry_model/ellipsoidal_chunk.cc]: This needs c++11. Can you use ```=Point<dim>();``` instead? (herer and 3 times below)

**Comment by bangerth** [comment]
On 10/08/2015 10:06 AM, MFraters wrote:
> The only reason I can think of that the normal vector is zero is that
> all the points of the face are at the same location.

I probably misunderstand, but are you suggesting that you have a face 
where every point is at the same location (and the face therefore has 
zero area)? This wouldn't make any sense, of course...

**Comment by tjhei** [comment]
call stack:
```
#0  0x00007ffff0755a3d in dealii::Tensor<0, 3, double>::operator/=<double> (this=0x7fffffff7500, s=0)
    at /ssd/deal-git/include/deal.II/base/tensor.h:716
#1  0x00007ffff07559ba in dealii::Tensor<1, 3, double>::operator/=<double> (this=0x7fffffff7500, s=0)
    at /ssd/deal-git/include/deal.II/base/tensor.h:1027
#2  0x00007ffff167fe80 in dealii::VectorTools::compute_nonzero_normal_flux_constraints<3, dealii::DoFHandler, 3> (dof_handler=..., 
    first_vector_component=0, boundary_ids=std::set with 5 elements, function_map=std::map with 5 elements, constraints=..., 
    mapping=...) at /ssd/deal-git/include/deal.II/numerics/vector_tools.templates.h:5546
Python Exception <type 'exceptions.IndexError'> list index out of range: 
#3  0x00007ffff16860de in dealii::VectorTools::compute_no_normal_flux_constraints<3, dealii::DoFHandler, 3> (dof_handler=..., 
    first_vector_component=0, boundary_ids=std::set with 5 elements, constraints=..., mapping=...)
    at /ssd/deal-git/include/deal.II/numerics/vector_tools.templates.h:5221
#4  0x000000000107939e in aspect::Simulator<3>::setup_dofs (this=0x7fffffff8220) at /ssd/aspect-git/source/simulator/core.cc:1142
#5  0x00000000010780b7 in aspect::Simulator<3>::run (this=0x7fffffff8220) at /ssd/aspect-git/source/simulator/core.cc:2023
#6  0x0000000000d99b85 in main (argc=2, argv=0x7fffffffc8f8) at /ssd/aspect-git/source/main.cc:510
```
code:
```
5541	            Tensor<1,dim> normal;
5542	            for (typename CellToNormalsMap::const_iterator
5543	                 x = cell_to_normals_map.begin();
5544	                 x != cell_to_normals_map.end(); ++x)
5545	              normal += x->second.first;
5546	            normal /= normal.norm();
```
values:
```
(gdb) print normal
$4 = {0, 0, 0}
(gdb) print cell_to_normals_map
$5 = std::map with 2 elements = {
  [{
  tria = 0x7fffffff8e18,
  dof_handler = 0x7fffffff9718,
  present_level = 1,
  present_index = 2
}] = {
    first = {0.17103064816215147, 0.037891820749926149, -0.98453680850920444}, 
    second = 1
  },
  [{
  tria = 0x7fffffff8e18,
  dof_handler = 0x7fffffff9718,
  present_level = 1,
  present_index = 8
}] = {
    first = {-0.17103064816215147, -0.037891820749926149, 0.98453680850920444}, 
    second = 1
  }
}
```

So two dofs contribute to a normal and the are ```n``` and ```-n```? That looks weird.

**Comment by MFraters** [comment]
@bangerth I was thinking of ways in which the normal vector could be zero, by thinking the setup of the a grid points/face might be wrong in the way that all the points are at the same location. But that is clearly not the case, especially with the results Timo shows.

**Comment by MFraters** [comment]
I now notice that there is something wrong with the boundary indicators, especially with the coordinate parallel one, where it only does the upper and lower one correct. So I am now looking what is going on there :)

**Comment by MFraters** [comment]
I noticed that the when I use the constant boundary temperature model, that the boundaries are 'almost correct'. As you can see in the figure below, on two boundaries there is a line of 0 values in the centre. 

part_shell

I am not sure if this is connected to the problem, but I noticed that is is quite easy to get an assert of cell_to_normals_map.size() == 1, when playing around with the boundaries. I have already tried to generate the parallelepiped with first the depth, but in the coordinate parallel version that resulted in similar problem and the non-coordinate parallel version crashed with the cell_to_normals_map.size() == 1 assert.

**Comment by tjhei** [comment]
Can you experiment with commenting out the code that does the transformation (push_forward and pull_back) and see how the output looks like and if everything works as expected?

**Comment by bangerth** [comment]
@MFraters -- was the picture generated by a parallel or a serial computation? My suspicion is that the boundary indicator of that one line is not correctly set. This may be because it is not correctly set on the coarse mesh, or because it is not correctly set upon parallel mesh refinement. The documentation of the `parallel::distributed::Triangulation` class discusses how to do this correctly upon parallel mesh refinement.

**Comment by MFraters** [comment]
@tjhei I tried that, but it resulted in crashes. Will look again into it tomorrow.

@bangerth It is a serial computation. I am seeing the problem both in the mesh with and without refinement. I am already doing something with the indicators, but I might do it incorrect. I will look at it.

**Comment by bangerth** [comment]
On 10/12/2015 04:04 PM, MFraters wrote:
> @bangerth <https://github.com/bangerth> It is a serial computation. I am
> seeing the problem both in the mesh with and without refinement. I am already
> doing something with the indicators, but I might do it incorrect. I will look
> at it.
I haven't looked at your code, so can only speculate, but are you calling 
`cell->set_boundary_indicator()` or `cell->set_all_boundary_indicators()`? Try 
understanding the differences between the two.

**Comment by MFraters** [comment]
After I create the grid and transform it i call 

```
for (Triangulation<3>::active_cell_iterator cell =
             coarse_grid.begin_active(); cell != coarse_grid.end(); ++cell)
        for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
          if (cell->face(f)->at_boundary())
#if DEAL_II_VERSION_GTE(8,3,0)
            cell->face(f)->set_all_boundary_ids(f);
#else
            cell->face(f)->set_all_boundary_indicators(f);
#endif
```

**Comment by bangerth** [comment]
This may or may not be correct (I honestly don't know). The calls to `set_all_boundary_indicators()` set boundary ids on edges, and calls on different faces may overwrite what the previous one has set for a common edge. 

Can you create a simple testcase where your mesh is uniformly refined and you have a boundary temperature model that assigns constant temperatures to each face? This should be relatively easy to do with the existing boundary temperature models and one should be able to look at the resulting temperature field at the end of the first time step to see what is happening.

**Comment by MFraters** [comment]
I think I have found something which causes the problem. To make the ellipsoidal chunk I now use the subdivided_parallelpiped. When I set the subdivision all to one (so it creates one cell), the problem goes away and I can subdivide using the initial global refinement without a problem. But when the subdivisions are not equal to one, the problem appears. I have noticed in the code of the subdivided parallelepiped that for the colorization it calls colorize_hyper_rectangle instead of colorize_subdivided_hyper_rectangle, which explain why it works for one initial cell, but not for a subdivided mesh.

**Comment by tjhei** [comment]
yes, subdivided_parallelpiped coloring is broken. I will look at fixing this in deal.II and get back to you.

**Comment by MFraters** [comment]
Thanks for looking at this!

**Comment by tjhei** [comment]
see https://github.com/dealii/dealii/pull/1750

To fix this we need to copy the implementation into compat.h if deal.II is older than 8.4, but we should wait until the feature got merged into deal.II in case it needs changes.

**Comment by MFraters** [comment]
I tried the fix in there and it solves at least the problem of that line. Now I hope it will also solve the FPE problem :)

**Comment by MFraters** [comment]
Do you agree that the relative differences of 1.1038343895e-6 and 2.8655840490e-6 are small enough to just apply the changes.diff to the test files?

**Comment by bangerth** [comment]
Yes, just go ahead with that.

**Comment by MFraters** [comment]
I went through the code again, performed some tests, and added some extra documentation. I also rebased it to the most current version of aspect because of the added documentation of the depth in the geometry model. So I think it is ready to merge. Is there anything else you want to see changed, added or tested?

**Comment by tjhei** [comment]
looks good to me, @bangerth ?

File: include/aspect/geometry_model/ellipsoidal_chunk.h [hunk 13]
```diff
@@ -56,11 +56,8 @@ namespace aspect
             set_initial_values(double para_semi_major_axis_a,
                                double para_eccentricity,
                                double para_semi_minor_axis_b,
-                               double para_rot_para_to_para_angle,
-                               double para_para_to_rect_angle,
-                               double para_rotation_longitude,
-                               double para_rotation_latitude,
-                               double para_bottom_depth);
+                               double para_bottom_depth,
+                               std::vector<std::vector<double> > para_corners);
```
Comment (bangerth) [inline include/aspect/geometry_model/ellipsoidal_chunk.h]: All of these are input arguments. Please mark them as `const`, and make the last argument a reference -- no need to copy a vector of points.

File: include/aspect/geometry_model/ellipsoidal_chunk.h [hunk 14]
```diff
@@ -198,31 +191,11 @@ namespace aspect
...
-        get_rotation_longitude() const;
-
-
-        /**
-         * Retrieve the rotation_latitude variable.
-         */
-        double
-        get_rotation_latitude() const;
+        std::vector<std::vector<double> >
+        get_corners() const;
```
Comment (bangerth) [inline include/aspect/geometry_model/ellipsoidal_chunk.h]: Here too, I think you'd be better off returning a `const` reference, rather than a copy.

Can you explain in the documentation of the function how "corners" are defined, and how that corresponds to the elements of the vector?

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 15]
```diff
@@ -65,20 +59,14 @@ namespace aspect
     EllipsoidalChunk<dim>::EllipsoidalChunkGeometry::set_initial_values(double para_semi_major_axis_a,
```
Comment (bangerth) [inline source/geometry_model/ellipsoidal_chunk.cc]: By the way, why is the function called `set_initial_values`? "Initial" means something at the beginning of time. Would `set_geometry_parameters` or similar be a better name?

File: include/aspect/geometry_model/ellipsoidal_chunk.h [hunk 16]
```diff
@@ -74,23 +71,12 @@ namespace aspect
...
           private:
             double semi_major_axis_a;
             double eccentricity;
             double semi_minor_axis_b;
-            double rot_para_to_para_angle;
-            double para_to_rect_angle;
-            double rotation_longitude;
-            double rotation_latitude;
             double bottom_depth;
+            std::vector<std::vector<double> > corners;
```
Comment (bangerth) [inline include/aspect/geometry_model/ellipsoidal_chunk.h]: I think it would be useful to document what the elements of this array correspond to. I suspect the coordinates of the corners of the model, but in which coordinate system? And how many elements is this array supposed to have? Also, would it be less confusing if you stored a `std::vector<Point<dim>>` here? This would make the semantics of the variable clearer.

**Comment by bangerth** [comment]
I had a few smaller comments. If you could address these, then I think it's ready to be merged.

File: include/aspect/geometry_model/ellipsoidal_chunk.h (hunk 16 above)
Comment (MFraters) [inline include/aspect/geometry_model/ellipsoidal_chunk.h]: The documentation is added now. About the type, I remember that there where some conversion problems internally. But I will look into that again.

**Comment by bangerth** [comment]
@MFraters -- let me know when you're done!

**Comment by MFraters** [comment]
I added a few assertTrows to make sure the new temp vector is the correct size (defined by what the user inputs), and tested different combinations of coordinates. So if the tester here still works, I think it is ready :)
